tests/e2e/hashkv_test.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2024 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"go.etcd.io/etcd/client/pkg/v3/fileutil"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+func TestVerifyHashKVAfterCompact(t *testing.T) {
+	scenarios := []struct {
+		clusterVersion e2e.ClusterVersion
+		keys           []string // used for data generators
+	}{
+		{
+			clusterVersion: e2e.CurrentVersion,
+			keys:           []string{"key0"},
+		},
+		{
+			clusterVersion: e2e.CurrentVersion,
+			keys:           []string{"key0", "key1"},
+		},
+		{
+			clusterVersion: e2e.QuorumLastVersion,
+			keys:           []string{"key0"},
+		},
+		{
+			clusterVersion: e2e.QuorumLastVersion,
+			keys:           []string{"key0", "key1"},
+		},
+	}
+
+	for _, compactedOnTombstoneRev := range []bool{false, true} {
+		for _, scenario := range scenarios {
+			t.Run(fmt.Sprintf("compactedOnTombstone=%v - %s - Keys=%v", compactedOnTombstoneRev, scenario.clusterVersion, scenario.keys), func(t *testing.T) {
+				e2e.BeforeTest(t)
+
+				if scenario.clusterVersion != e2e.CurrentVersion {
+					if !fileutil.Exist(e2e.BinPath.EtcdLastRelease) {
+						t.Skipf("%q does not exist", e2e.BinPath.EtcdLastRelease)
+					}
+				}
+
+				ctx := context.Background()
+
+				cfg := e2e.NewConfigClientTLS()
+				clus, err := e2e.NewEtcdProcessCluster(ctx, t,
+					e2e.WithConfig(cfg),
+					e2e.WithClusterSize(3),
+					e2e.WithVersion(scenario.clusterVersion))
+				require.NoError(t, err)
+
+				t.Cleanup(func() { clus.Close() })
+
+				tombstoneRevs, lastestRev := populateDataForHashKV(t, clus, cfg.Client, scenario.keys)
+
+				compactedOnRev := tombstoneRevs[0]
+
+				// If compaction revision is not tombstone, select revision after 3 writes from first tombstone.
+				// And ensure it's not the following tombstone.
+				const afterWriters = int64(3)
+				if !compactedOnTombstoneRev {
+					compactedOnRev += afterWriters
+					require.True(t, tombstoneRevs[1] > compactedOnRev)
+				}
```

Comment:
The `3` may cause unnecessary confusion. We don't care about the number at all, we just need to ensure the compaction revision isn't a tombstone.
```suggestion
				// If compaction revision isn't a tombstone, select a revision in the middle of two tombstones.
				if !compactedOnTombstoneRev {
					compactedOnRev = (tombstoneRevs[0] + tombstoneRevs[1]) / 2
					require.True(t, tombstoneRevs[0] < compactedOnRev && compactedOnRev < tombstoneRevs[1])
				}				
```

Reply:
Updated.